Update GridViewRenderer to appropriately handle change of ItemsSource

pfaucon opened the pull request:
Current GridViewRenderer does not handle a change of item source elegantly, and instead ignores any change to its binding.  I changed the naming because it looks like there may have been confusion form the original writer (or from me, although inverting the +/- worked for me).  The DataCollectionChanged force-call is required once, otherwise there won't be any updates until the collection changes again (insert/delete)

Additionally ReloadData is a bit heavy if people are working with large collectionViews, I haven't made any changes there but it might be nice if people are modifying larger data sets (not rebinding)

ravensorb commented:
Is there a test case for this?

pfaucon commented:
I'm not sure exactly how to add a test case for this, if you could advise I would be happy to take a look at it though(I don't find any examples in the iOS unit testing).

The expected behavior is:
- Attach a new ObservableCollection as an items source
- Modify the observable collection
- The DataCollectionChanged should be called 

In my manual tests it was not, but it was called after the sign inversion.

The other commit with modifications to the actual dataCollectionChanged I'm also not sure how to test.  Can we inject a different Control in for the Control property and verify the changes?

ravensorb commented:
Maybe a change to the sample app that allows for randomly generating new items in the ItemSource (or adding a new one on button click?)

pfaucon commented:
That makes sense, [here(1)](https://github.com/pfaucon/XamarinFormsGridExample) is a basic project showing the two examples; the one from nuget, and the custom one copied from my proposed changes (with the namespace changed to match the project).

You won't notice a loading lag when adding items because the cells are very simple and light, but in our use case where the gridview is full of camera images (stored as byte[] and converted when shown the reload is much more obvious). If I get time I'll update it, but think I it would be enough to show that the converter is being called multiple times whenever a new item is added (without the proposed modification).

(1) https://github.com/pfaucon/XamarinFormsGridExample

ravensorb commented:
Looks good.  Do you want to take that idea and add it into the Sample App?

Here is a link to the GridViewPage in the sample app
https://github.com/XLabs/Xamarin-Forms-Labs/blob/master/Samples/XLabs.Sample/Pages/Controls/GridViewPage.xaml

pfaucon commented:
Sorry, yeah, that would have made a lot more sense to do directly.  I added the changes with some commits to the patch-1 branch.  

I think the reason that the original developer heavy-handedly did a reloadData was that calls that are made asynchronously to the UI are basically ignored, so my approach actually didn't work directly in the sample app, but I modified it to queue a block on the made thread before inserting indexes (lightweight so it shouldn't be a problem).

In the renderer there is a try-catch block, I don't think that it belongs there (in my experience nothing inside should be in a situation to throw exceptions), but I don't feel comfortable modifying that.

ravensorb commented:
Nice!  Now all we need is a Windows implementation of this control.

pfaucon commented:
I would be willing but I don't have a win10 vm, and my windows 8 vm won't update, apparently virtual machines aren't eligible for the update :)

ravensorb commented:
Annoying -- I though as long as it wasn't the enterprise edition it qualified :-/  

BTW -- I was thinking maybe we use the ItemsControl as the underlying control for Windows.  Thoughts?
